File: src/load-prg.c
/*----------------------------------------------------------------------
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------

 *-------------------------- load-prg.c -------------------------------------
 * The functions in this file implement decoding algorithms for the .PRG 
 * file format. It includes, for exemple, code to extract the background image
 * of a pinball table.
 *----------------------------------------------------------------------
 */

#include <gtk/gtk.h>
#include "prg.h"
#define IMAGE_XMAX 320

//------------------Prototypes--------------------------
void load_prg(char* filename, struct prg_contents *table);
void load_palette(FILE *fp, struct prg_contents *table);
void decode_bgimage(FILE *fp, struct prg_contents *table);
void decode_masks(FILE *fp, struct prg_contents *table);
void decode_image(FILE *fp, unsigned long int file_offset, unsigned char* image, int xsize, int ysize);
void decode_mask(FILE *fp, char* mask, int xsize, int ysize, unsigned int addr);
extern void parsexml(char *docname, struct prg_contents* table);
extern char* new_filename_change_extension(char*,char*);
//------------------------------------------------------

void load_prg(char* filename, struct prg_contents *table){
	FILE *fp;
	int i,j;
	char *conf_file = new_filename_change_extension(filename,".xml");

	parsexml(conf_file, table);
	free(conf_file);

	if ((fp = fopen(filename,"r"))<0){
                g_print("Error while trying to open %s\n", filename);
                exit(-1);
        }

	//strcopy (table->filename, filename);
	i=0;
	while((table->filename[i] = filename[i]) != '\0')
		i++;

	decode_bgimage(fp, table);
	decode_masks(fp, table);
	load_palette(fp, table);
	fclose(fp);
}

void decode_image(FILE *fp, unsigned long int file_offset, unsigned char* image, int xsize, int ysize){
	int xplot=0, yplot=0, i;
	unsigned char count,color;

        fseek(fp,file_offset,SEEK_SET);
	printf("image: %x-",ftell(fp));
	
        while(yplot < ysize){
		count = getc(fp);
		if (count <= 0x80){
                        count++;
			while(count--){
				image[xsize*yplot + xplot++] = getc(fp);
				if (xplot >= xsize){
					xplot=0;
					yplot++;
				}
			}
		} else {
			color = getc(fp);
			for(i=0;i<257-count;i++){
				image[xsize*yplot + xplot++] = color;
				if (xplot >= xsize){
					xplot=0;
					yplot++;
				}
			}
		}
        }
	printf("%x\n",ftell(fp)-1);
}



void decode_bgimage(FILE *fp, struct prg_contents *table){
	int i;
	for(i=0;i<4;i++){
		decode_image(fp, table->bgi_addr[i], &(table->bgimage[i*320*144]), 320, 144);
	}
}




void decode_mask(FILE *fp, char* mask, int xsize, int ysize, unsigned int addr){	
	unsigned int xplot=1,yplot=1,j;
	unsigned char byte;
	
	fseek(fp,addr,SEEK_SET);
	printf("%x-",ftell(fp));
	while(yplot < ysize){
		byte = getc(fp);
		for(j=128;j>0;j/=2){
			mask[xsize*(yplot-1) + xplot++ -1] = (byte & j)/j;
			if (xplot > xsize){
				xplot=1;
				yplot++;
			}
		}
	}
	
}

void decode_masks(FILE *fp, struct prg_contents *table){	
	int i;
	
        for(i=0;i<8;i++){
		printf("mask%d: ",i);
		decode_mask(fp, table->mask[i], 320, 576, table->mask_addr[i]);
		printf("%x\n",ftell(fp)-1);
	}
	printf("flipper: ");
	decode_mask(fp, table->flipper, 64, 576, table->flipper_addr);
	printf("%x\n",ftell(fp)-1);
}

void load_palette(FILE *fp, struct prg_contents *table){
	int i;
	fseek(fp, table->palette_addr, SEEK_SET);
	printf("Palette: %x-", ftell(fp));
	for (i=0;i<256;i++){
		table->palette[i].red = getc(fp);		
		table->palette[i].green = getc(fp);
		table->palette[i].blue = getc(fp);
	}
	printf("%x\n", ftell(fp)-1);
}

File: pfantedit-py/loadprg.py
#!/usr/bin/env python

import string
import array

def palette(fp, addr):
	fp.seek(addr)
	status = fp.tell()
	pal = []
	for i in range(256):
		red = ord(fp.read(1))
		green = ord(fp.read(1))
		blue = ord(fp.read(1))
		pal.append( (red*256 + green)*256 + blue )
	print "Palette: ", status, "-", fp.tell(), "\n"
	return pal

code={}
def decodeimage(fp, file_offset, xsize, ysize, debug=False):
	global code
	image = array.array('B', " " * xsize * ysize)
#	if debug == True:
	if True:
		fp.seek(file_offset)
		code[file_offset]=[]
		for i in range(xsize*ysize):
			code[file_offset].append(ord(fp.read(1)))
		#print "encoded: "
		#print code[file_offset]

#	print "original code keys:", code.keys()
	
	fp.seek(file_offset)
	#states definition:
	START = 1
	COMPRESSED = 2
	RAW = 3
	
	state = START
	index = 0
	while index < xsize*ysize:
		if state == START:
			byte = ord(fp.read(1))
			if byte<=0x80:
				state = RAW
				count = byte + 1
			else:
				state=COMPRESSED
				count = 257 - byte
				color = ord(fp.read(1))
		elif state == COMPRESSED:
			if count == 0:
				state = START
			else:
				count = count - 1	
				image[index]=color
				index = index + 1
		elif state == RAW:
			if count == 0:
				state = START
			else:
				image[index] = ord(fp.read(1))
				index = index + 1
				count = count - 1
	print "image: ", file_offset, "-", fp.tell()
	if debug == True:
		print "raw :"
		print image

	return image

def decode_mask(fp, xsize, ysize, addr):
	mask = array.array('B', " " * xsize * ysize)
	xplot = 0
	yplot = 0
	fp.seek(addr)
	while(yplot < ysize):
		byte = ord(fp.read(1))
		j=128
		while j>0:
			mask[xsize*(yplot-1) + xplot -1] = (byte & j)/j
			xplot = xplot + 1
			if (xplot > xsize):
				xplot = 1
				yplot = yplot + 1
			j = j / 2
	print "mask: ", addr, "-", fp.tell()
	return mask

def load_sensors(fp, addr, n):
	fp.seek(addr)
	sensors = []
	a = array.array('B', " " * 8)
	while n>0:
		a[0]=ord(fp.read(1))
		a[1]=ord(fp.read(1))
		a[2]=ord(fp.read(1))
		a[3]=ord(fp.read(1))
		a[4]=ord(fp.read(1))
		a[5]=ord(fp.read(1))
		a[6]=ord(fp.read(1))
		a[7]=ord(fp.read(1))

		while a[1] > 1 or a[3] > 1 or a[5] > 1 or a[7] > 1:
			for i in range(6):
				a[i]=a[i+2]
			a[6]=ord(fp.read(1))
			a[7]=ord(fp.read(1))
		n=n-1
		sensors.append([a[0]+a[1]*256, a[2]+a[3]*256, a[4]+a[5]*256, a[6]+a[7]*256])
				
	return sensors
